hs/src/Reach/Eval/Core.hs

```diff
@@ -3122,6 +3122,20 @@ evalPrim p sargs =
       bs <- mustBeBytes bsv
       vm <- evalAsEnv (lvl, obj)
       retV $ (lvl, SLV_Bool at $ M.member (bunpack bs) vm)
+    SLPrim_Object_set -> do
+      at <- withAt id
+      (obj, fieldV, valV) <- three_args
+      (objTy, objDLA) <- compileTypeOf obj
+      (valTy, valDLA) <- compileTypeOf valV
+      fieldName <- bunpack <$> mustBeBytes fieldV
+      case objTy of
+        T_Object objFields -> do
+          let objFields' = M.insert fieldName valTy objFields
+          let objTy' = T_Object objFields'
+          let dle = DLE_ObjectSet at objDLA fieldName valDLA
```

Comment:
You need to enforce that the type is the same. You'd do this by checking that `objTy` already has `fieldName` with the type `valTy`